libia2/include/ia2.h

```diff
@@ -264,13 +254,94 @@ static int insecure_pkey_mprotect(void *ptr, size_t len, int prot, int pkey) {
+      printf("invalid thread id\n");                                           \
+      exit(1);                                                                 \
+    }                                                                          \
+    size_t stack_idx = (i * PAGE_SIZE / sizeof(void *)) + tid;                 \
+    char *stack = allocate_stack(i);                                           \
+    protect_stack(i, stack);                                                   \
+                                                                               \
+    /* Each stack frame start + 8 is initially 16-byte aligned. */             \
+    char *stack_ptr_value = stack + STACK_SIZE - 8;                            \
+                                                                               \
+    /* We must change the pkru to write the stack pointer because each page */ \
+    /* of ia2_stackptrs is pkey_mprotected to the compartment whose stacks */  \
+    /* it contains. */                                                         \
+                                                                               \
+    /* Save the current pkru. */                                               \
+    uint32_t old_pkru = 0;                                                     \
+    __asm__ volatile(IA2_RDPKRU : "=a"(old_pkru) : "a"(0), "d"(0), "c"(0));    \
+    /* Switch to the compartment's pkru to access its stack pointer. */        \
+    uint32_t new_pkru = ~((3 << (2 * i)) | 3);                                 \
+    __asm__ volatile(IA2_WRPKRU : : "a"(new_pkru), "d"(0), "c"(0));            \
+    /* Forbid overwriting an existing stack. */                                \
+    if (ia2_stackptrs[stack_idx] != NULL) {                                    \
+      printf(                                                                  \
+          "compartment %zd in thread %zd tried to allocate existing stack\n",  \
+          i, tid);                                                             \
+      exit(1);                                                                 \
+    }                                                                          \
+    /* Write the stack pointer. */                                             \
+    ia2_stackptrs[stack_idx] = stack_ptr_value;                                \
+    /* Restore the old pkru. */                                                \
+    __asm__(IA2_WRPKRU : : "a"(old_pkru), "d"(0), "c"(0));                     \
+  }                                                                            \
+                                                                               \
+  size_t ia2_alloc_compacted_tid(void);                                        \
+                                                                               \
+  void init_stacks(void) {                                                     \
```

Comment:
The definition of `init_stacks` needs to statically depend (via the macro argument `n`) on the number of compartments that exist in the compartmentalized program, so it needs to be defined inside the `INIT_RUNTIME` (actually `INIT_RUNTIME_COMMON`) macro in the `ia2.h` header. It ends up in the main compartment's library rather than linked in via libia2.

Reply:
Got it, alright.